File: code/java/ch16/src/main/java/com/github/zxh0/luago/state/Closure.java
package com.github.zxh0.luago.state;

import com.github.zxh0.luago.api.JavaFunction;
import com.github.zxh0.luago.binchunk.Prototype;
import lombok.Getter;

@Getter
class Closure {

    final Prototype proto;
    final JavaFunction javaFunc;
    final UpvalueHolder[] upvals;

    // Lua Closure
    Closure(Prototype proto) {
        this.proto = proto;
        this.javaFunc = null;
        this.upvals = new UpvalueHolder[proto.getUpvalues().length];
    }

    Closure(JavaFunction javaFunc, int nUpvals) {
        this.proto = null;
        this.javaFunc = javaFunc;
        this.upvals = new UpvalueHolder[nUpvals];
    }

}

File: code/csharp/ch09/luavm/state/Closure.cs
using luavm.api;

namespace luavm.state
{
    public struct Closure
    {
        internal binchunk.Prototype proto;
        internal CsharpFunction csharpFunc;

        internal static Closure newLuaClosure(binchunk.Prototype proto)
        {
            return new Closure
            {
                proto = proto
            };
        }

        internal static Closure newCsharpClosure(CsharpFunction charpFunc)
        {
            return new Closure
            {
                csharpFunc = charpFunc
            };
        }
    }
}